eng/packaging.targets

```diff
@@ -202,6 +202,18 @@
     </ItemGroup>
   </Target>
 
+  <Target Name="WarnOnProjectReferenceToFrameworkAssemblies"
+          BeforeTargets="IncludeTransitiveProjectReferences"
+          Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)' AND
+                     '@(ProjectReference)' != ''">
+    <Warning Text="Project reference '%(ProjectReference.Filename)' is a reference to a framework assembly and is not required."
+             Code="NETPKG0001"
+             Condition="$(NetCoreAppLibrary.Contains('%(ProjectReference.Filename);')) AND
```

Comment:
Would it make sense to source this from the list of shared framework assemblies that is returned by an SDK target instead of our own custom property / item? That would make this target more portable, i.e. when being used with other shared frameworks (aspnetcore, windowsdesktop).

Reply:
The real customer solution is to warn folks when direct references are dropped by conflict resolution.  That works for everyone, not just us folks building frameworks.

I considered using the targeting pack as the source here but felt it would complicate the solution and hurt perf.  I'll give it a try and share the sample that does this as well as the perf difference.

Reply:
Another short-coming of using the ref-pack output is that it doesn't work for things within the repo that are part of the shared framework (like System.Text.Json) since those don't resolve the framework.   As a result, we'd still need to use the list for those.

Reply:
> Another short-coming of using the ref-pack output is that it doesn't work for things within the repo that are part of the shared framework (like System.Text.Json) since those don't resolve the framework. As a result, we'd still need to use the list for those.

Agreed. That's a good reason for using the static NetCoreAppLibrary.props list as the data source. Consider using the [NetCoreAppLibrary item](https://github.com/dotnet/runtime/blob/070fbe41b1adc81093c610ec310fa13c317a826c/src/libraries/NetCoreAppLibrary.props#L249) instead of the property for better perf (i.e. via an item function).

The property is necessary to use it as a condition for other properties but here in a target, the item is already available.

Reply:
Here's the alternative solution. https://github.com/ericstj/runtime/commit/ad97b9e61ab24aa39655ee7134dc3ccf8aaa2487

Perf with this alternative solution (`build libs -allConfigurations`)
      475 ms  WarnOnProjectReferenceToFrameworkAssemblies 205 calls

Perf with the solution in this PR (`build libs -allConfigurations`)
       38 ms  WarnOnProjectReferenceToFrameworkAssemblies 205 calls

  

It's quite clunky and we still need to pull in the static list.

Reply:
> That's a good reason for using the static NetCoreAppLibrary.props list as the data source. Consider using the [NetCoreAppLibrary item](https://github.com/dotnet/runtime/blob/070fbe41b1adc81093c610ec310fa13c317a826c/src/libraries/NetCoreAppLibrary.props#L249) instead of the property for better perf

Cool I didn't know we had that.  I'll make that change.

Reply:
It's actually slower to use the item:
`      407 ms  WarnOnProjectReferenceToFrameworkAssemblies 205 calls`

This is because when using the item I cannot run a single check within the batching that produces the warning message.  Instead, I need to do an intersection.  I tried a few different solutions - intersecting with condition, intersecting with include/exclude, using item functions and target batching but and nothing brought the time down in the realm of my first solution.  I think I'll stick with that.    We already use it in a similar fashion here:
https://github.com/dotnet/runtime/blob/8fe777f51b3c1eff82ebbce1495f889305f9eec9/eng/references.targets#L42

I agree that hypothetically MSBuild should be able to give me something faster here with an item check - it ought to have a hash-set for the NetCoreAppLibrary and ought to let me look up into that set based on the filename of ProjectReference (which should be faster than a string search) - but I cannot find a way to do that.  cc @rainersigwald  in case he sees this and can share a pattern that's better.